### PR TITLE
BE#42-Admin-API-Endpoints

### PR DESCRIPTION
**Pull request by SoySauceLabs**
# TOpreDOs:
- [ ] `source env_vars.txt` from **Project Documentation > Wiki+Reference > BE > env_vars.txt**
# TOtestDOs:
- [ ] Reference the `/admin` endpoints in the API document from **Project Documentation > Wiki+Reference > API >Basics**
- [ ] Ask Matt for test objects if the document is too enigmatic.
# Itinerary
- [ ] Sip it.

### CODE REVIEW

File: src/admin/views.py [hunk 1]
```diff
@@ -0,0 +1,517 @@
...
+    """
+
+    if request.method != 'POST':
+        return HttpResponseNotAllowed(['POST'])
+
+    data = json.loads(request.body.decode('utf-8'))
+
+    auth_token = str(data.get('auth_token', ''))
+    admin_username = str(data.get('admin_username', ''))
+    user_edit_info = data.get('edit_user_info', '')
```
Comment (scotte216) [inline src/admin/views.py]: user_edit_info vs edit_user_info... Seems like these should have the same name.

File: src/cassy.py [hunk 2]
```diff
@@ -557,3 +557,553 @@ def change_user_email(username, new_email):
...
+    parameters = {
+            'username': new_user_info.get('username', ''),
+            'password': new_user_info.get('password', ''),
+            'email': new_user_info.get('email', ''),
+            'admin': new_user_info.get('admin', ''),
+            'enable': new_user_info.get('enable', ''),
+            'securitytoken': new_user_info.get('securitytoken', ''),
+            'subenddate': new_user_info.get('subenddate', ''),
+            'userid': int(new_user_info.get('userid', '')),
+            'vineyards': new_user_info.get('vineyards', ''),
```
Comment (scotte216) [inline src/cassy.py:652]: Is there a mechanism that does some kind of checking about these parameters? Like that userids are positive integers w/o duplicate (or whatever), sub end date is actually a date in the future, etc.

File: src/cassy.py [hunk 3]
```diff
@@ -557,3 +557,553 @@ def change_user_email(username, new_email):
...
+            'password': rows[0].password,
+            'admin': rows[0].admin,
+            'email': rows[0].email,
+            'enable': rows[0].enable,
+            'securitytoken': rows[0].securitytoken,
+            'subenddate': rows[0].subenddate,
+            'userid': rows[0].userid,
+            'vineyards': rows[0].vineyards,
+        }
+        if (user_edit_info.get('username', '') != ''):
```
Comment (scotte216) [inline src/cassy.py]: Is there a way to do something like:
for each key in user_edit_info: 
    if user_edit_info.get(key, '') != ''):
        edit_row[key] = user_edit_info.get(key, '')

**Comment by scotte216** [comment]
Did a code review. Still need to do manual testing.

File: src/admin/views.py (hunk 1 above)
Comment (SoySauceLabs) [inline src/admin/views.py]: fixed!

File: src/cassy.py (hunk 2 above)
Comment (SoySauceLabs) [inline src/cassy.py:652]: Good catch! There is not currently, but I'll see if I can add some reasonable checks in.

File: src/cassy.py (hunk 3 above)
Comment (SoySauceLabs) [inline src/cassy.py]: Yes, actually I believe that would work quite well here, I'll give it a touch up!

**Comment by scotte216** [comment]
endpoint admin/user/new 
Using an existing username from the DB but with a different pwd:

Expected behavior: tell us the username is bad / already exists.

Actual behavior: will accept duplicate usernames as long as the passwords are different.

**Comment by scotte216** [comment]
endpoint admin/user/new

Create a new user with junk for the subenddate

Expected behavior: Error about bad time/date

Actual behavior: 500 -- Unknown error occurred 

NOTE: admin/user/subscription returns a different error when the date is bad:
"unknown": "An unknown error occurred. time data '2017-o1-31' does not match format '%Y-%m-%d'"

**Comment by scotte216** [comment]
endpoint admin/user/new
Not sure if this is a problem but -- I'm allowed to set the subenddate to a date in the past.

**Comment by SoySauceLabs** [comment]
All righty, I've addressed constraining the usernames to be unique, and fixed the subscription date checking for correct format and "futurama-ness" of the date! 📆

**Comment by scotte216** [comment]
When creating a new user with username "" I get the following error response

"sub_end_date_invalid": "Subscription end date is invalid. username_invalid"

The subscription date IS valid, but the username is invalid.

**Comment by scotte216** [comment]
The username is allowed to be a single space: " " 
I didn't want to make a ton of such users, but I imagine it's just reading spaces as another character. Probably want to trim whitespace.

**Comment by scotte216** [comment]
I'm being pickier with the user info since it seems like the kind of thing where that's most likely to be screwed up.

**Comment by scotte216** [comment]
/admin/vineyard
With a vineyard ID of a non-existent vineyard (20), I got the following response:
 "vineyard_bad_id": "A vineyard ID must be a positive integer."

Expected:
'The vineyard ID was not found.'

**Comment by scotte216** [comment]
vineyard/new
Allowed create of vineyard with vineyard "name" attribute missing.

API doc says it's required.

Also allowed creation of a new vineyard with missing boundary attribute.

**Comment by scotte216** [comment]
vineyard/edit

Incorrect error when using a non-existent vineyard ID:
Expected: The vineyard ID was not found
Actual: "A vineyard ID must be a positive integer."

**Comment by scotte216** [comment]
vineyard/

Strange issue:
Any vineyard ID greater than 2 results in the following error:
"auth_error_not_found": "Auth token not found."

File: src/admin/views.py [hunk 4]
```diff
@@ -0,0 +1,691 @@
...
+
+    data = json.loads(request.body.decode('utf-8'))
+
+    auth_token = str(data.get('auth_token', ''))
+    admin_username = str(data.get('admin_username', ''))
+    request_username = str(data.get('request_username', ''))
+
+    try:
+        is_verified = verify_admin(admin_username, auth_token)
+        if (is_verified is False):
```
Comment (scotte216) [inline src/admin/views.py]: Could simplify these two lines
if not verify_admint(admin_username, auth_token): 

OR 

if not is_verified:

**Comment by scotte216** [comment]
General comment/observation:

We have the ability to disable a user and disable a vineyard. To enable a user you can use the /user/edit endpoint. According to the API documentation, the vineyard/edit endpoint doesn't seem to allow for re-enabling. 

Also, displaying user (admin/user) or vineyard information (admin/vineyard) doesn't show whether they are en-/disabled or not.

**Comment by SoySauceLabs** [comment]
Sweet! Okay, so I've added a substantially more robust set of parameter validations and plenty of new unit tests. I believe I've addressed _all_ of the above comments. In addition, the API doc should be up to date with missing/invalid information!

File: src/admin/views.py [hunk 5]
```diff
@@ -29,22 +30,22 @@
 logger = logging.getLogger('plantalytics_backend.admin')
 
 
-def verify_admin(admin_username, auth_token):
+def verify_admin(auth_token):
     """
     Verifies username has admin privileges and is authenticated.
```
Comment (scotte216) [inline src/admin/views.py:33]: Function description is inaccurate.

**Comment by scotte216** [comment]
It's not worth holding up the PR for a method description comment being inaccurate.

**Comment by scotte216** [comment]
So actually... When I create a new user and the subscription end-date has already passed, the log gets it right, but the returned error code is incorrect.

"unknown": "An unknown error occurred."

**Pull request merged**